### ISSUE TITLE
Omitting typings for `this` when function signatures are unioned

### ISSUE THREAD

**Issue by 19majkel94** [Microsoft/TypeScript#27862]
Looks like the TS compiler loose types info about `this` for function when the type is an union. See code below.

**TypeScript Version:**  3.2.0-dev.201xxxxx

**Search Terms:**
this union
this callback

**Code**

```ts
type NoParamCallback = (this: number) => void;
type ParamCallback = (this: number, param: any) => void;

type NoParamCallbackTest = (callback: NoParamCallback) => void;
const noParamCallbackTest = (() => {}) as NoParamCallbackTest;
noParamCallbackTest(function() {
  const num: number = this;
});

type NoOrParamCallbackTest = (
  callback: NoParamCallback | ParamCallback
) => void;
const noOrParamCallbackTest = (() => {}) as NoOrParamCallbackTest;
noOrParamCallbackTest(function() {
  const num: number = this;
});
```

**Expected behavior:**
Compile without error. Type of `this` in callback is `number`

**Actual behavior:**
```
[ts] 'this' implicitly has type 'any' because it does not have a type annotation.
```

**Playground Link:** 
http://www.typescriptlang.org/play/#src=type%20NoParamCallback%20%3D%20(this%3A%20number)%20%3D>%20void%3B%0D%0Atype%20ParamCallback%20%3D%20(this%3A%20number%2C%20param%3A%20any)%20%3D>%20void%3B%0D%0A%0D%0Atype%20NoParamCallbackTest%20%3D%20(callback%3A%20NoParamCallback)%20%3D>%20void%3B%0D%0Aconst%20noParamCallbackTest%20%3D%20(()%20%3D>%20%7B%7D)%20as%20NoParamCallbackTest%3B%0D%0AnoParamCallbackTest(function()%20%7B%0D%0A%20%20const%20num%3A%20number%20%3D%20this%3B%0D%0A%7D)%3B%0D%0A%0D%0Atype%20NoOrParamCallbackTest%20%3D%20(%0D%0A%20%20callback%3A%20NoParamCallback%20%7C%20ParamCallback%0D%0A)%20%3D>%20void%3B%0D%0Aconst%20noOrParamCallbackTest%20%3D%20(()%20%3D>%20%7B%7D)%20as%20NoOrParamCallbackTest%3B%0D%0AnoOrParamCallbackTest(function()%20%7B%0D%0A%20%20const%20num%3A%20number%20%3D%20this%3B%0D%0A%7D)%3B

**Related Issues:** 
Haven't found

**Comment by andy-ms** [comment]
This applies to unions in general:
```ts
type Cb0 = (param: number) => void;
type Cb1 = (param: number, other: string) => void;
declare function f(callback: Cb0 | Cb1): void;
f(p => {}); // Errors currently
```
Think this is a duplicate of #7294.

A workaround is to figure out the union signature yourself -- in this case `callback: (this: number, param?: any) => void`.

**Comment by 19majkel94** [comment]
> A workaround is to figure out the union signature yourself -- in this case callback: (this: number, param: any) => void.

Yes, this is a simplified case - in reality it has also different return signature, like:

```ts
type One = (first: number) => void;
type Two = (first: number, second: number) => number;
```

The only workaround I've found for now is using manual union and inlinined conditional types:
```ts
type OneOrTwo = (first: number, second: number) => number | void

declare function f<T extends OneOrTwo>(callback: T):
  T extends One
  ? ReturnType<One>
  : (T extends Two
    ? ReturnType<Two>
    : never);

const one: void = f((first) => { });
const two: number = f((first, second) => second);
const three: void = f(() => {});
```

But it's not an elegant solution 😕
